File: fitness_core/bodyfat.py
"""Body-fat estimation formulas.

Sources:
- Siri (1961) two-compartment equation.
- US Navy circumference method (metric form).
- Jackson & Pollock 3-site (1978/1980) body density + Siri.
- Deurenberg (1991) BMI-based.
"""
import math
from fitness_core.common import Sex


def siri(density: float) -> float:
    """Siri (1961) two-compartment equation. density in g/cm^3 -> body-fat %."""
    return 495 / density - 450


def navy_bf(sex: Sex, waist_cm: float, neck_cm: float, height_cm: float,
            hip_cm: float | None = None) -> float:
    log10 = math.log10
    if sex == Sex.MALE:
        if waist_cm <= neck_cm:
            raise ValueError(
                f"waist_cm ({waist_cm}) must be greater than neck_cm ({neck_cm})")
        d = (1.0324 - 0.19077 * log10(waist_cm - neck_cm)
             + 0.15456 * log10(height_cm))
    else:
        if hip_cm is None:
            raise ValueError("hip_cm is required for female Navy estimate")
        if waist_cm + hip_cm <= neck_cm:
            raise ValueError(
                f"waist_cm + hip_cm ({waist_cm + hip_cm}) must be greater than "
                f"neck_cm ({neck_cm})")
        d = (1.29579 - 0.35004 * log10(waist_cm + hip_cm - neck_cm)
             + 0.22100 * log10(height_cm))
    return siri(d)


def jackson_pollock_3(sex: Sex, sum_mm: float, age: float) -> float:
    if sex == Sex.MALE:
        density = (1.10938 - 0.0008267 * sum_mm + 0.0000016 * sum_mm ** 2
                   - 0.0002574 * age)
    else:
        density = (1.0994921 - 0.0009929 * sum_mm + 0.0000023 * sum_mm ** 2
                   - 0.0001392 * age)
    return siri(density)


def deurenberg(sex: Sex, bmi: float, age: float) -> float:
    sexf = 1 if sex == Sex.MALE else 0
    return 1.20 * bmi + 0.23 * age - 10.8 * sexf - 5.4

File: packages/core/src/tools/body-fat.ts
import { z } from "zod";
import { Sex } from "../math/common.js";
import { LengthSchema, MassSchema, massKg, lengthCm } from "../math/units.js";
import { navyBf, jacksonPollock3, deurenberg } from "../math/bodyfat.js";
import { computeConsensus, roundTo } from "../math/stats.js";
import { DomainError } from "../errors.js";
import {
  MethodResultSchema, SkippedMethodSchema, ConsensusSchema,
} from "../models.js";
import { MethodsSchema, resolveMethods, runMethods, type MethodOutput } from "../dispatch.js";
import type { Tool } from "../registry.js";

const ALL_METHODS = ["navy", "jackson-pollock-3", "deurenberg"];
const REASONS: Record<string, string> = {
  navy: "navy: requires neck, waist, height (+ hip for female)",
  "jackson-pollock-3": "jackson-pollock-3: requires skinfold_sum and age",
  deurenberg: "deurenberg: requires weight, height, age",
};

export const BodyFatInput = z.object({
  sex: Sex,
  age: z.number().gt(0).lte(120).nullable().default(null),
  height: LengthSchema.nullable().default(null),
  weight: MassSchema.nullable().default(null),
  neck: LengthSchema.nullable().default(null),
  waist: LengthSchema.nullable().default(null),
  hip: LengthSchema.nullable().default(null),
  skinfold_sum: z.number().gt(0).nullable().default(null),
  methods: MethodsSchema,
});
export type BodyFatInputT = z.output<typeof BodyFatInput>;

export const BodyFatOutput = z.object({
  results: z.array(MethodResultSchema),
  consensus: ConsensusSchema.nullable(),
  skipped: z.array(SkippedMethodSchema),
});
export type BodyFatOutputT = z.output<typeof BodyFatOutput>;

function detail(inp: BodyFatInputT, bf: number): Record<string, unknown> | null {
  if (inp.weight === null) return null;
  const kg = massKg(inp.weight);
  const fat = roundTo((kg * bf) / 100, 2);
  return { fat_mass_kg: fat, lean_mass_kg: roundTo(kg - fat, 2) };
}

function bfValue(method: string, inp: BodyFatInputT): number | null {
  if (method === "navy") {
    if (inp.neck && inp.waist && inp.height) {
      if (inp.sex === "female" && inp.hip === null) return null;
      try {
        return navyBf(
          inp.sex,
          lengthCm(inp.waist),
          lengthCm(inp.neck),
          lengthCm(inp.height),
          inp.hip ? lengthCm(inp.hip) : undefined,
        );
      } catch (e) {
        throw new DomainError((e as Error).message);
      }
    }
    return null;
  }
  if (method === "jackson-pollock-3") {
    if (inp.skinfold_sum !== null && inp.age !== null)
      return jacksonPollock3(inp.sex, inp.skinfold_sum, inp.age);
    return null;
  }
  if (method === "deurenberg") {
    if (inp.weight && inp.height && inp.age !== null) {
      const bmi = massKg(inp.weight) / (lengthCm(inp.height) / 100) ** 2;
      return deurenberg(inp.sex, bmi, inp.age);
    }
    return null;
  }
  throw new DomainError(`unknown method: ${method}`);
}

export function compute(inp: BodyFatInputT): BodyFatOutputT {
  const run = (method: string): MethodOutput => {
    const bf = bfValue(method, inp);
    if (bf === null) return null;
    return [bf, detail(inp, bf)];
  };
  const { requested, explicit } = resolveMethods(inp.methods, ALL_METHODS);
  const { results, skipped } = runMethods(requested, explicit, run, "%", {
    reasonFn: (m) => REASONS[m] ?? `${m}: required inputs missing`,
    ndigits: 2,
  });
  return { results, consensus: computeConsensus(results.map((r) => r.value)), skipped };
}

export const tool: Tool<BodyFatInputT, BodyFatOutputT> = {
  id: "body-fat",
  name: "Body Fat Percentage",
  description:
    "Estimate body-fat % via US Navy circumference, Jackson-Pollock 3-site skinfold, " +
    "and Deurenberg (BMI-based) methods.",
  category: "composition",
  tags: ["body fat", "composition", "navy", "skinfold"],
  methods: ALL_METHODS,
  input: BodyFatInput,
  output: BodyFatOutput,
  compute,
  examples: [
    {
      input: {
        sex: "male",
        neck: { value: 40, unit: "cm" },
        waist: { value: 90, unit: "cm" },
        height: { value: 180, unit: "cm" },
      },
      output: { results: [{ method: "navy", unit: "%" }] },
    },
  ],
};
